build.sc

```diff
@@ -57,13 +57,38 @@ trait PPrintMainModule extends CrossScalaModule {
           )
         """
     }
+
+    val twrapOutput = s"""
+        case class TWrap[T](shouldWrap: Boolean) {
+          def getWrapped(s: String) = if (shouldWrap) s"($$s)" else s
+        }
+        trait TWrapLowerPriorityImplicits {
+          implicit def nonFunctionsAreNotWrapped[T]: TWrap[T] = new TWrap[T](false)
+        }
+        object TWrap extends TWrapLowerPriorityImplicits {
+          private type L = Nothing
+          private type U = Any
+          implicit def functions0AreWrapped[T](implicit ev: T <:< Function0[U]): TWrap[T] = new TWrap[T](true)
+          ${
+            1.to(22)
+              .map(i => s"""
+          implicit def functions${i}AreWrapped[T](implicit ev: T <:< Function$i[${1.to(i).map(_ => "L").mkString(",")}, U]): TWrap[T] = new TWrap[T](true)"""
+              )
+              .mkString
+          }
+        }
+    """
+
     val output = s"""
         package pprint
+
+        $twrapOutput
+
         trait TPrintGen[Type[_], Cfg]{
           def make[T](f: Cfg => String): Type[T]
-          def get[T: Type](cfg: Cfg): String
+          def get[T: Type: TWrap](cfg: Cfg): String
           implicit def F0TPrint[R: Type]: Type[() => R] = make[() => R](cfg => "() => " + get[R](cfg))
```

Comment:
Removed the `TWrap` constraint on type `R` (force-pushed on this PR's commit), which was causing right hand side to be wrapped if function, which is not necessary. I also added relevant test case: `pprint.TWrapTests.higherKindedFunctions.lazy`.